docs/index.rst

```diff
@@ -1,63 +1,179 @@
+start experimenting with pluggy functionality.
 
-In fact, ``pytest`` is itself composed as a set of ``pluggy`` plugins
-which are invoked in sequence according to a well defined set of protocols.
-Some `200+ plugins`_ use ``pluggy`` to extend and customize ``pytest``'s default behaviour.
+.. literalinclude:: examples/toy-example.py
 
-In essence, ``pluggy`` enables function `hooking`_ so you can build "pluggable" systems.
+Running this directly gets us::
 
-How's it work?
---------------
-A `plugin` is a `namespace`_ which defines hook functions.
+    $ python docs/examples/toy-example.py
+
+    inside Plugin_2.myhook()
+    inside Plugin_1.myhook()
+    [-1, 3]
 
-``pluggy`` manages *plugins* by relying on:
+A complete example
+------------------
+Now let us demonstrate how this plays together in a vaguely real world scenario.
 
-- a hook *specification* - defines a call signature
-- a set of hook *implementations* - aka `callbacks`_
-- the hook *caller* - a call loop which collects results
+Let's assume our ``host program`` is called **eggsample** where some eggs will
+be prepared and served with a tray containing condiments. As everybody knows:
+the more cooks are involved the better the food, so let us make the process
+pluggable and write a plugin that improves the meal with some spam and removes
+the steak sauce from the condiments tray (nobody likes that anyway).
 
-where for each registered hook *specification*, a hook *call* will invoke up to ``N``
-registered hook *implementations*.
+.. note::
 
-``pluggy`` accomplishes all this by implementing a `request-response pattern`_ using *function*
-subscriptions and can be thought of and used as a rudimentary busless `publish-subscribe`_
-event system.
+    **naming markers**: ``HookSpecMarker`` and ``HookImplMarker`` must be
+    the initialized with the name of the ``host`` project (the ``name``
+    parameter in ``setup()``) - so **eggsample** in our case.
 
-``pluggy``'s approach is meant to let a designer think carefully about which objects are
-explicitly needed by an extension writer. This is in contrast to subclass-based extension
-systems which may expose unnecessary state and behaviour or encourage `tight coupling`_
-in overlying frameworks.
+    **naming plugin projects**: they should be named in the form of
+    ``<host>-<plugin>`` (e.g. ``pytest-xdist``), therefore we call our
+    plugin *eggsample-spam*.
 
+The host
+^^^^^^^^
+``eggsample/eggsample/__init__.py``
 
-A first example
----------------
+.. literalinclude:: examples/eggsample/eggsample/__init__.py
 
-.. literalinclude:: examples/firstexample.py
+``eggsample/eggsample/hookspecs.py``
 
-Running this directly gets us::
+.. literalinclude:: examples/eggsample/eggsample/hookspecs.py
 
-    $ python docs/examples/firstexample.py
+``eggsample/eggsample/lib.py``
 
-    inside Plugin_2.myhook()
-    inside Plugin_1.myhook()
-    [-1, 3]
+.. literalinclude:: examples/eggsample/eggsample/lib.py
```

Comment:
I personally like including line numbers with [`:linenos:`](http://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html?highlight=literalinclude#directive-literalinclude).

Reply:
Added it ... and removed it again. Looks absoulutely horrific due to https://github.com/sphinx-doc/sphinx/issues/2427 and does not really add too much utility IMO.

Reply:
Meh, not that big of a deal. Yeah just avoid the headache.